### PR TITLE
Switch from Travis CI to GitHub Actions

### LINKED ISSUE #4124: Travis CI free plan limit reached, switch CI/CD provider?
We did it, we reached the limits of the free plan from Travis CI :tada: :confused: As a result, new commits (such as accdc072778769907fb27dee4c5f3a029218856a), including those to PRs, and new PRs won't build any more.

Though I think it is unlikely we will reach the limit very often - I think we reached it this month due to to amazing amount of contributions by @fbernhart :heart: - I do still suggest we look into alternatives. Even if it is just to make sure we have builds this last week of November.

I'd personally suggest we start using [GitHub Actions](https://github.com/features/actions) for everything (we already [use it for a couple of things](https://github.com/simple-icons/simple-icons/tree/develop/.github/workflows)) since it is free for open source projects. If there are no objections I will open a PR for the switch later today.

In the meantime, @simple-icons/maintainers be extra careful with what you merge/commit!

### PR DESCRIPTION
**Pull request by ericcornelissen**
This replaces the Travis CI configuration file by (hopefully) equivalent GitHub Actions Workflows. The result is two GitHub Actions Workflows, one that lints, tests, and builds on pushes and Pull Requests, and one that deploys to NPM and creates a release tag on pushes to `master`. ~~Note that I did not yet test the latter.~~

I now (1 day after opening) tested - and fixed - the publishing workflow as well. An example of a working run is [this one](https://github.com/ericcornelissen/simple-icons/actions/runs/382865128). You can see that the NPM deployment (should) work in [these logs](https://github.com/ericcornelissen/simple-icons/runs/1452632770#step:6:167) and that the GitHub release is created in [these logs](https://github.com/ericcornelissen/simple-icons/runs/1452632826#step:10:73) - and [here is the release on my fork](https://github.com/ericcornelissen/simple-icons/releases/tag/3.13.0).

The GitHub release title is `Release {{ VERSION }}` and the GitHub release text is taken directly from the commit on which the job is running (and our [release-action](https://github.com/simple-icons/release-action) ensures that this is the list of new/edited/removed icons). Note that this slightly changes the release title, currently it is formatted roughly as `Release {{ NEW_ICONS }} and {{ UPDATED_ICONS }} and {{ REMOVED_ICONS }}`. If we want that format back we can obtain it by looking at the commit title of the commit that triggers the job since our [release-action](https://github.com/simple-icons/release-action) ensures the commit title has that format.

~~Things that still need to be looked in to:~~

- [x] Git clone depth (similar to #3847)
- [x] Caching (similar to #3847)

Other things to do:

- [x] Add `NPM_TOKEN` secret to repository
- ~~Add `GITHUB_USERNAME` secret to repository~~ No longer required
- ~~Add `GITHUB_EMAIL` secret to repository~~ No longer required
- [x] Update [the build badge](https://github.com/simple-icons/simple-icons#status) in the README

* * *

Closes #4124. But, as per the discussion there, we might still decide to stick with Travis if we get the free plan for OSS projects.

### CODE REVIEW

**Comment by mondeja** [comment]
> Git clone depth (similar to #3847).

In [actions/checkout README](https://github.com/actions/checkout#checkout-v2) it's specified that "only a single commit is fetched by default, for the ref/SHA that triggered the workflow." So TravisCI `git.depth: 1` is the default behaviour in `actions/checkout`.

File: .github/workflows/verify.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Use Ruby 2.5
+      uses: actions/setup-ruby@v1
+      with:
+        ruby-version: 2.5
+    - name: Install Jekyll
+      run: gem install jekyll
+    - name: Build website
+      run: jekyll build
```
Comment (mondeja) [inline .github/workflows/verify.yml:36]: > Caching (similar to #3847)

This do the trick:

```suggestion
    - name: Setup cache for gems
      uses: actions/cache@v2
      with:
        path: vendor/bundle
        key: ${{ runner.os }}-gems-${{ hashFiles('**/Gemfile.lock') }}
        restore-keys: |
          ${{ runner.os }}-gems-
    - name: Install Jekyll
      run: |
        echo -e 'source "https://rubygems.org"\n\ngem "jekyll"' > Gemfile
        bundle config set path 'vendor/bundle'
        bundle install --jobs 4 --retry 3
    - name: Build website
      run: bundle exec jekyll build
```

**Review by mondeja — CHANGES_REQUESTED** [review]

**Review by mondeja — APPROVED** [review]

**Comment by ericcornelissen** [comment]
> In [actions/checkout README](https://github.com/actions/checkout#checkout-v2) it's specified that "only a single commit is fetched by default, for the ref/SHA that triggered the workflow." So TravisCI `git.depth: 1` is the default behaviour of `actions/checkout`.
> 
> This behaviour is [shown in the logs](https://github.com/simple-icons/simple-icons/pull/4126/checks?check_run_id=1450097901#step:2:30).

Thanks for verifying this, I remember something like this but I didn't have time left to check it myself.


> - Hash all `Gemfile.lock` files of gems installed in `vendor/bundle` as key for cache.
> - Create a simple `Gemfile` to install with bundler as dependencies for the project.
> - Install the "jekyll" gem into `vendor/bundle`.
> - Execute jekyll with `bundle exec` to make sure that the locally installed jekyll is used.

And thanks for the explanation + code suggestions 😃 I committed them directly as you suggested. That took the "Build website" job time down from 3m 31s to 14s! 🎉 

Just for the record [given this part of the actions/cache documentation](https://github.com/actions/cache/blob/0781355a23dac32fd3bac414512f4b903437991a/examples.md#node---npm) I'd say we shouldn't try to cache things in the NodeJS based jobs.

**Comment by ericcornelissen** [comment]
I have now tested - and fixed - the publishing workflow as well. An example of a working run is [this one](https://github.com/ericcornelissen/simple-icons/actions/runs/382865128). You can see that the NPM deployment (should) work in [these logs](https://github.com/ericcornelissen/simple-icons/runs/1452632770#step:6:167) and that the GitHub release is created in [these logs](https://github.com/ericcornelissen/simple-icons/runs/1452632826#step:10:73) - and [here is the release on my fork](https://github.com/ericcornelissen/simple-icons/releases/tag/3.13.0).

The GitHub release title is `Release {{ VERSION }}` and the GitHub release text is taken directly from the commit on which the job is running (and our [release-action](https://github.com/simple-icons/release-action) ensures that this is the list of new/edited/removed icons). Note that this slightly changes the release title, currently it is formatted roughly as `Release {{ NEW_ICONS }} and {{ UPDATED_ICONS }} and {{ REMOVED_ICONS }}`. If we want that format back we can obtain it by looking at the commit title of the commit that triggers the job since our [release-action](https://github.com/simple-icons/release-action) ensures the commit title has that format. Thoughts on this @simple-icons/maintainers?

**Comment by mondeja** [comment]
What about include the current title in the release text and let the title as `Release {{ VERSION }}`? Like:

## v3.13.0

```
Release 1 new icon and 29 updated icons.

# New Icons

- Numba (#3791)

# Updated Icons

- About.me (#3883)
- Airbnb (#3874)
- Angular (#3814)
- Arch Linux (#3808)
- Buy Me A Coffee (#3836)
- eBay (#3875)
- Eclipse IDE (#3803)
- Etsy (#3876)
- Facebook (#3856)
- Gatsby (#3848)
- Gentoo (#3810)
- GNOME (#3811)
- Hulu (#3832)
- iTunes (#3833)
- jQuery (#3815)
- Laravel (#3849)
- Lumen (#3850)
- Messenger (#3857)
- Myspace (#3858)
- Pepsi (#3878)
- PowerShell (#3804)
- PyTorch (#3819)
- QuickTime (#3834)
- Rust (#3820)
- TikTok (#3859)
- Twitter (#3860)
- Udemy (#3877)
- Vine (#3835)
- Xcode (#3806)
```

**Comment by ericcornelissen** [comment]
To clarify, the title of current releases is actually `Release {{ NEW_ICONS }} and {{ UPDATED_ICONS }} and {{ REMOVED_ICONS }} ({{ VERSION }})` (as seen, e.g. in [the latest release](https://github.com/simple-icons/simple-icons/releases/tag/3.13.0)).

Also, given the current configuration the release would look like what @mondeja posted. Personally prefer the current approach, I think the release title being just the version number or `Release {{ VERSION }}` is a bit weird, especially in the context of the GitHub UI. But I'm fine either way. If can add a relatively simple step that extracts the first line from `steps.commit.outputs.git-message` and outputs the commit body separately from the title and use them accordingly in the `Create release` step. It doesn't seem [there exists any existing Action that gives it to you](https://github.com/marketplace?type=actions&query=commit+title)

**Comment by mondeja** [comment]
> Also, given the current configuration the release would look like what @mondeja posted.

Ah, I see :eyes: 

> Personally prefer the current approach, I think the release title being just the version number or `Release {{ VERSION }}` is a bit weird...

For me it is fine both ways, as you prefer :+1:

**Review by service-paradis — APPROVED** [review]

**Comment by ericcornelissen** [comment]
> For me it is fine both ways, as you prefer +1

I've updated it for now to the way existing releases are formatted, just for consistency.

With that I will leave this open until A) [I hear back from Travis CI](https://github.com/simple-icons/simple-icons/issues/4124#issuecomment-733089991), or B) this Friday/weekend before the planned release of [v4](https://github.com/simple-icons/simple-icons/milestone/3).

**Comment by fbernhart** [comment]
Concerning GitHub Actions:

I just saw that if you click on one of the actions here (like: https://github.com/simple-icons/simple-icons/actions/runs/384016136) there is now a warning:

> Ubuntu-latest workflows will use Ubuntu-20.04 soon. For more details, see https://github.com/actions/virtual-environments/issues/1816

So Github Actions is changing from Ubuntu 18.04 to 20.04 for `runs-on: ubuntu-latest` and therefore there are as well some software related changes.

Looking at the two different Github Actions of simple-icons that are using `runs-on: ubuntu-latest` ("Verify" (this pull request) as well as "Pull Request Labeler") I would say that this doesn't affect simple-icons. But can someone with more experience double check?

Maybe you've already seen this warning - then feel free to ignore my comment. ☺️

**Comment by ericcornelissen** [comment]
Thanks for your comment @fbernhart, I already saw the warning and checked the issue you linked to. It seems the main difference is some of the software that comes preinstalled, but since we always explicitly install a specific NodeJS or Ruby version that is not a problem. Potentially something might go wrong with regards to the standard CLI tools (such as `cat` or `head`) installed in 20.04 vs 18.04, but I tested everything locally on 19.04 and it worked. I don't expect these simple and basic tools to change in 20.04 and if they do we can just install them manually.

From my point of view, I prefer to automatically stay up-to-date with the latest version of Ubuntu in the CI rather than having to update this manually. Especially since we are not doing anything fancy, I don't expect anything to go wrong with that any time soon.


> With that I will leave this open until A) [I hear back from Travis CI](https://github.com/simple-icons/simple-icons/issues/4124#issuecomment-733089991), or B) this Friday/weekend before the planned release of [v4](https://github.com/simple-icons/simple-icons/milestone/3).

As I haven't heard back from Travis CI yet (and I'm not expecting them to respond in the weekend), and with the approval of @mondeja and @service-paradis, I'm going to go ahead and merge this, finalizing the switch from Travis CI to GitHub Actions. This will allow us to safely merge Pull Requests again before the first of December and for the next release to be released by GitHub Actions.

**Comment by ericcornelissen** [comment]
> Potentially something might go wrong with regards to the standard CLI tools (such as cat or head) installed in 20.04 vs 18.04, but I tested everything locally on 19.04 and it worked.

Relatedly, [the step where the release title and body are extracted from the commit message](https://github.com/simple-icons/simple-icons/pull/4126/files#diff-551d1fcf87f78cc3bc18a7b332a4dc5d8773a512062df881c5aba28a6f5c48d7R47) may be removed if https://github.com/kceb/git-message-action/pull/6 gets merged. If not, alternatively, someone could release a new GitHub Actions action based on that and we could use that instead.

**Pull request merged**